sycl/include/CL/sycl/xilinx/fpga/kernel_properties.hpp

```diff
@@ -21,35 +21,166 @@
+};
+
+template <typename CharT, CharT... Sep, CharT... P1, CharT... P2,
+          typename... Rest>
+struct concat<cstr<CharT, Sep...>, cstr<CharT, P1...>, cstr<CharT, P2...>,
+              Rest...> {
+  using type =
+      typename concat<cstr<CharT, Sep...>, cstr<CharT, P1..., Sep..., P2...>,
+                      Rest...>::type;
+};
+
+#ifdef __SYCL_DEVICE_ONLY__
+
 #pragma clang diagnostic push
 #pragma clang diagnostic ignored "-Wgnu-string-literal-operator-template"
 
+/// utility to transform a string literal into its cstr form.
 template <typename CharT, CharT... charpack>
-__SYCL_ALWAYS_INLINE constexpr void operator"" _kernel_param() noexcept {
-  __SYCL_DEVICE_ANNOTATE("xilinx_kernel_param", FixedString<charpack...>)
-  int useless;
-  (void)useless;
+__SYCL_ALWAYS_INLINE constexpr cstr<CharT, charpack...>
+operator"" _cstr() noexcept {
+  return {};
 }
 #pragma clang diagnostic pop
+
+using Base2 = decltype("01"_cstr);
+using Base10 = decltype("0123456789"_cstr);
+using Base16 = decltype("0123456789abcdef"_cstr);
+
 #else
-__SYCL_ALWAYS_INLINE constexpr void
-operator"" _kernel_param(const char *, size_t) noexcept {}
+__SYCL_ALWAYS_INLINE constexpr cstr<char> operator"" _cstr(const char *,
+                                                           size_t) noexcept {
+  return {};
+}
+
+/// When not in device mode the kernel properties are not used so they do not
+/// need to be correct just to compile correctly.
+using Base2 = cstr<char, '0'>;
+using Base10 = cstr<char, '0'>;
+using Base16 = cstr<char, '0'>;
+
 #endif
 
+#ifdef __SYCL_DEVICE_ONLY__
+/// utility to transform a number constant like 348 into 
+/// cstr<char, '3', '4', '8'> such that it can be manipulated a a string.
+template<bool, typename>
+struct StrNumber {};
+#else
+template<bool, typename>
+struct StrNumber {
+  using type = cstr<char>;
+};
+#endif
+
+template <auto N, typename Base = Base10> struct number {
+  static constexpr auto value = N;
+  static constexpr auto str = typename StrNumber<(N >= Base::size), number>::type{};
+};
+
+#ifdef __SYCL_DEVICE_ONLY__
+template <auto N, typename Base> struct StrNumber<true, number<N, Base>> {
+  using last_didgit = typename Base::template at<N % Base::size>;
+  using type = typename concat<
+      cstr<char>,
+      typename StrNumber<(N / Base::size >= Base::size), number<N / Base::size, Base>>::type,
+      last_didgit>::type;
+};
+
+template <auto N, typename Base> struct StrNumber<false, number<N, Base>> {
+  using type = typename Base::template at<N % 10>;
+};
+#endif
+
+template <typename Func, typename... CharPacks>
+auto kernel_param_impl(Func &&f, CharPacks...) {
+  return sycl::detail::make_const_cheater(
+      [internal_f = std::forward<Func>(f)](auto &&... args) mutable {
+        __SYCL_DEVICE_ANNOTATE(
+            "xilinx_kernel_param",
+            sycl::detail::StrPacker<typename sycl::detail::concat<
+                sycl::detail::cstr<char, ' '>, CharPacks...>::type>::Str)
+        int useless;
```

Comment:
Comment...

Reply:
So I still do not know why you need this useless variable since there is no comment...